--- a/distrobox-enter
+++ b/distrobox-enter
@@ -332,7 +332,7 @@
 		if [ "${headless}" -eq 0 ]; then
 			container_command_login="${container_command_login} --pty"
 		fi
-		container_command_login="${container_command_login} -c \"\\\$(getent passwd ${USER} | cut -f 7 -d :) -l\""
+		container_command_login="${container_command_login} -c \"\\\$SHELL -l\""
 		if [ -n "${container_command}" ]; then
 			container_command="$(echo "${container_command}" | sed 's/\$/\\\$/g')"
 			container_command="${container_command_login} -c \"${container_command}\""
