diff --git a/kernel/common/inc/nv-time.h b/kernel/common/inc/nv-time.h
index 3d14f9bd..73fd880a 100644
--- a/kernel/common/inc/nv-time.h
+++ b/kernel/common/inc/nv-time.h
@@ -82,7 +82,7 @@ static inline NV_STATUS nv_sleep_us(unsigned int us)
     ktime_get_raw_ts64(&tm1);
 #endif
 
-    if (in_irq() && (us > NV_MAX_ISR_DELAY_US))
+    if (in_hardirq() && (us > NV_MAX_ISR_DELAY_US))
         return NV_ERR_GENERIC;
 
     mdelay_safe_msec = us / 1000;
@@ -127,7 +127,7 @@ static inline NV_STATUS nv_sleep_ms(unsigned int ms)
     tm_start = tm_aux;
 #endif
 
-    if (in_irq() && (ms > NV_MAX_ISR_DELAY_MS))
+    if (in_hardirq() && (ms > NV_MAX_ISR_DELAY_MS))
     {
         return NV_ERR_GENERIC;
     }
diff --git a/kernel/conftest.sh b/kernel/conftest.sh
index cfa38712..8a7384d5 100755
--- a/kernel/conftest.sh
+++ b/kernel/conftest.sh
@@ -1591,7 +1591,7 @@ compile_test() {
                 drm_atomic_helper_legacy_gamma_set();
             }"
 
-            compile_check_conftest "$CODE" "NV_DRM_ATOMIC_HELPER_LEGACY_GAMMA_SET_PRESENT" "" "functions"
+            #compile_check_conftest "$CODE" "NV_DRM_ATOMIC_HELPER_LEGACY_GAMMA_SET_PRESENT" "" "functions"
         ;;
 
         drm_plane_create_color_properties)
@@ -2371,7 +2371,7 @@ compile_test() {
                 drm_mode_connector_attach_encoder();
             }"
 
-            compile_check_conftest "$CODE" "NV_DRM_CONNECTOR_FUNCS_HAVE_MODE_IN_NAME" "" "functions"
+            #compile_check_conftest "$CODE" "NV_DRM_CONNECTOR_FUNCS_HAVE_MODE_IN_NAME" "" "functions"
         ;;
 
         drm_connector_has_vrr_capable_property)
@@ -3075,10 +3075,12 @@ compile_test() {
 
             if [ -f conftest$$.o ]; then
                 rm -f conftest$$.o
-                echo "#define NV_DRM_CRTC_ATOMIC_CHECK_HAS_ATOMIC_STATE_ARG" | append_conftest "types"
-            else
-                echo "#undef NV_DRM_CRTC_ATOMIC_CHECK_HAS_ATOMIC_STATE_ARG" | append_conftest "types"
+                #echo "#define NV_DRM_CRTC_ATOMIC_CHECK_HAS_ATOMIC_STATE_ARG" | append_conftest "types"
+            #else
+                #echo "#undef NV_DRM_CRTC_ATOMIC_CHECK_HAS_ATOMIC_STATE_ARG" | append_conftest "types"
             fi
+
+            echo "#define NV_DRM_CRTC_ATOMIC_CHECK_HAS_ATOMIC_STATE_ARG" | append_conftest "types"
         ;;
 
         drm_gem_object_vmap_has_map_arg)
@@ -3251,10 +3253,12 @@ compile_test() {
 
             if [ -f conftest$$.o ]; then
                 rm -f conftest$$.o
-                echo "#define NV_DRM_PLANE_ATOMIC_CHECK_HAS_ATOMIC_STATE_ARG" | append_conftest "types"
-            else
-                echo "#undef NV_DRM_PLANE_ATOMIC_CHECK_HAS_ATOMIC_STATE_ARG" | append_conftest "types"
+                #echo "#define NV_DRM_PLANE_ATOMIC_CHECK_HAS_ATOMIC_STATE_ARG" | append_conftest "types"
+            #else
+                #echo "#undef NV_DRM_PLANE_ATOMIC_CHECK_HAS_ATOMIC_STATE_ARG" | append_conftest "types"
             fi
+
+            echo "#define NV_DRM_PLANE_ATOMIC_CHECK_HAS_ATOMIC_STATE_ARG" | append_conftest "types"
         ;;
 
         ib_peer_memory_symbols)
@@ -4416,7 +4420,8 @@ compile_test() {
                 return 0;
             }"
 
-            compile_check_conftest "$CODE" "NV_DRM_MODE_CREATE_DP_COLORSPACE_PROPERTY_HAS_SUPPORTED_COLORSPACES_ARG" "" "types"
+            #compile_check_conftest "$CODE" "NV_DRM_MODE_CREATE_DP_COLORSPACE_PROPERTY_HAS_SUPPORTED_COLORSPACES_ARG" "" "types"
+            echo "#define NV_DRM_MODE_CREATE_DP_COLORSPACE_PROPERTY_HAS_SUPPORTED_COLORSPACES_ARG" | append_conftest "types"
         ;;
 
         drm_syncobj_features_present)
@@ -4803,10 +4808,12 @@ compile_test() {
             if [ -f conftest$$.o ]; then
                 rm -f conftest$$.o
 
-                echo "#define NV_ASSIGN_STR_ARGUMENT_COUNT 1" | append_conftest "functions"
-            else
-                echo "#define NV_ASSIGN_STR_ARGUMENT_COUNT 2" | append_conftest "functions"
+                #echo "#define NV_ASSIGN_STR_ARGUMENT_COUNT 1" | append_conftest "functions"
+            #else
+                #echo "#define NV_ASSIGN_STR_ARGUMENT_COUNT 2" | append_conftest "functions"
             fi
+
+            echo "#define NV_ASSIGN_STR_ARGUMENT_COUNT 1" | append_conftest "functions"
         ;;
 
         drm_driver_has_date)
@@ -4853,7 +4860,8 @@ compile_test() {
                 .mode_valid = conftest_drm_connector_mode_valid,
             };"
 
-            compile_check_conftest "$CODE" "NV_DRM_CONNECTOR_HELPER_FUNCS_MODE_VALID_HAS_CONST_MODE_ARG" "" "types"
+            #compile_check_conftest "$CODE" "NV_DRM_CONNECTOR_HELPER_FUNCS_MODE_VALID_HAS_CONST_MODE_ARG" "" "types"
+            echo "#define NV_DRM_CONNECTOR_HELPER_FUNCS_MODE_VALID_HAS_CONST_MODE_ARG" | append_conftest "types"
         ;;
 
         memory_device_coherent_present)
diff --git a/kernel/nvidia/nv-backlight.c b/kernel/nvidia/nv-backlight.c
index a4f2d040..8cf3010a 100644
--- a/kernel/nvidia/nv-backlight.c
+++ b/kernel/nvidia/nv-backlight.c
@@ -38,7 +38,7 @@ NV_STATUS NV_API_CALL nv_get_tegra_brightness_level
     nv_linux_state_t *nvl = NV_GET_NVL_FROM_NV_STATE(nv);
     struct backlight_device *bd;
 
-    bd = get_backlight_device_by_name(nvl->backlight.device_name);
+    bd = backlight_device_get_by_name(nvl->backlight.device_name);
     if (bd == NULL)
     {
         nv_printf(NV_DBG_ERRORS, "Unable to get backlight device\n");
@@ -63,7 +63,7 @@ NV_STATUS NV_API_CALL nv_set_tegra_brightness_level
     nv_linux_state_t *nvl = NV_GET_NVL_FROM_NV_STATE(nv);
     struct backlight_device *bd;
 
-    bd = get_backlight_device_by_name(nvl->backlight.device_name);
+    bd = backlight_device_get_by_name(nvl->backlight.device_name);
     if (bd == NULL)
     {
         nv_printf(NV_DBG_ERRORS, "Unable to get backlight device\n");
diff --git a/kernel/nvidia/os-interface.c b/kernel/nvidia/os-interface.c
index 835e329d..bb3243de 100644
--- a/kernel/nvidia/os-interface.c
+++ b/kernel/nvidia/os-interface.c
@@ -371,7 +371,7 @@ NvBool NV_API_CALL os_semaphore_may_sleep(void)
 
 NvBool NV_API_CALL os_is_isr(void)
 {
-    return (in_irq());
+    return (in_hardirq());
 }
 
 // return TRUE if the caller is the super-user
diff --git a/kernel/common/inc/nv-mm.h b/kernel/common/inc/nv-mm.h
index faeb8445..c1d6b7a7 100644
--- a/kernel/common/inc/nv-mm.h
+++ b/kernel/common/inc/nv-mm.h
@@ -202,7 +202,7 @@ static inline void nv_vm_flags_set(struct vm_area_struct *vma, vm_flags_t flags)
 {
 #if !NV_CAN_CALL_VMA_START_WRITE
     nv_vma_start_write(vma);
-    ACCESS_PRIVATE(vma, __vm_flags) |= flags;
+    vma_flags_set_word(&vma->flags, flags);
 #elif defined(NV_VM_AREA_STRUCT_HAS_CONST_VM_FLAGS)
     vm_flags_set(vma, flags);
 #else
@@ -214,7 +214,7 @@ static inline void nv_vm_flags_clear(struct vm_area_struct *vma, vm_flags_t flag
 {
 #if !NV_CAN_CALL_VMA_START_WRITE
     nv_vma_start_write(vma);
-    ACCESS_PRIVATE(vma, __vm_flags) &= ~flags;
+    vma_flags_clear_word(&vma->flags, flags);
 #elif defined(NV_VM_AREA_STRUCT_HAS_CONST_VM_FLAGS)
     vm_flags_clear(vma, flags);
 #else
diff --git a/kernel/nvidia-drm/nvidia-drm-helper.h b/kernel/nvidia-drm/nvidia-drm-helper.h
index a297b943..c340156d 100644
--- a/kernel/nvidia-drm/nvidia-drm-helper.h
+++ b/kernel/nvidia-drm/nvidia-drm-helper.h
@@ -196,7 +196,7 @@ int nv_drm_atomic_helper_disable_all(struct drm_device *dev,
        for ((__i) = 0;                                                      \
             (__i) < (__state)->num_connector &&                             \
             ((connector) = (__state)->connectors[__i].ptr,                  \
-            (connector_state) = (__state)->connectors[__i].state, 1);       \
+            (connector_state) = (__state)->connectors[__i].state_to_destroy, 1);       \
             (__i)++)                                                        \
                for_each_if (connector)
 #else
@@ -223,7 +223,7 @@ int nv_drm_atomic_helper_disable_all(struct drm_device *dev,
        for ((__i) = 0;                                                \
             (__i) < (__state)->dev->mode_config.num_crtc &&           \
             ((crtc) = (__state)->crtcs[__i].ptr,                      \
-            (crtc_state) = (__state)->crtcs[__i].state, 1);           \
+            (crtc_state) = (__state)->crtcs[__i].state_to_destroy, 1);           \
             (__i)++)                                                  \
                for_each_if (crtc_state)
 #else
@@ -248,7 +248,7 @@ int nv_drm_atomic_helper_disable_all(struct drm_device *dev,
        for ((__i) = 0;                                                   \
             (__i) < (__state)->dev->mode_config.num_total_plane &&       \
             ((plane) = (__state)->planes[__i].ptr,                       \
-            (plane_state) = (__state)->planes[__i].state, 1);            \
+            (plane_state) = (__state)->planes[__i].state_to_destroy, 1);            \
             (__i)++)                                                     \
                for_each_if (plane_state)
 #else
diff --git a/kernel/nvidia-drm/nvidia-drm-priv.h b/kernel/nvidia-drm/nvidia-drm-priv.h
index 703725c3..9aa19d73 100644
--- a/kernel/nvidia-drm/nvidia-drm-priv.h
+++ b/kernel/nvidia-drm/nvidia-drm-priv.h
@@ -33,6 +33,7 @@
 
 #include <drm/drm_device.h>
 #include <drm/drm_gem.h>
+#include <drm/drm_print.h>
 
 #include "nvidia-drm-os-interface.h"
 
